File: leetcode/merge-intervals.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals = sorted(intervals)

        length = len(intervals)

        res = []
        start = 0
        end = 0

        for i in range(1, length):
            if intervals[i][0] <= intervals[end][1]:
                if intervals[i][1] > intervals[end][1]:
                    end = i
                continue

            res.append([intervals[start][0], intervals[end][1]])
            start = i
            end = i
        res.append([intervals[start][0], intervals[end][1]])
        return res

File: leetcode/merge-intervals.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<vector<int>> merge(vector<vector<int>>& intervals)
    {
        vector<vector<int>> res;

        auto size = intervals.size();

        sort(intervals.begin(), intervals.end());

        for (auto interval : intervals) {
            if (res.empty() || res.back()[1] < interval[0]) {
                res.push_back(interval);
            } else {
                res.back()[1] = max(res.back()[1], interval[1]);
            }
        }

        return res;
    }
};
